### PR TITLE
Enable block signature verification

### LINKED ISSUE #834: Signing blocks
We should pursue signing blocks. Just signing inputs in PoS is not enough as some components (like, say, block rewards) can be mutated.

### LINKED ISSUE #875: Block signature signature verification
Based on #874 Two open questions:

1. The block verification should be done, in line with proof of stake.
2. Do we want to include the public key of the signature? This has the pro that we can verify the header (and hence the block) easily and add more trust, given that the same key that signs the stake kernel is the key signing the header. Hence, if that's done, then the only verification we need is that the output is valid, which is doable in other ways.

### PR DESCRIPTION
**Pull request by TheQuantumPhysicist**
Closes #834
Closes #875

### CODE REVIEW

File: chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                        // Spending an output of a pool creation transaction is only allowed in a
+                        // context of a transaction (as opposed to block reward) only if this pool
+                        // is being decommissioned.
+                        // If this rule is being invalidated, it will be detected in other parts
+                        // of the code.
+                        Ok(pool_data.decommission_key().clone())
+                    }
+                    TxOutput::ProduceBlockFromStake(_, pool_id) => {
+                        // The only way we can spend this output in a transaction
+                        // (as opposed to block reward), is if we're decommissioning a mempool
```
Comment (azarovh) [inline chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs:81]: typo: a pool

**Review by azarovh — APPROVED** [review]

File: chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::chain::{Destination, PoolId, TxOutput};
+use pos_accounting::PoSAccountingView;
+
+use super::accounting_delta_adapter::PoSAccountingDeltaAdapter;
+
+pub type SignatureDestinationGetterFn<'a> =
+    dyn Fn(&TxOutput) -> Result<Destination, SignatureDestinationGetterError> + 'a;
```
Comment (iljakuklic) [inline chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs:22]: Seems to me it should be possible to return a reference to destination here, it is always cloned in the implementation below as far as I can see. The lifetimes may be non-trivial though so likely not worth it.

**Review by iljakuklic — APPROVED** [review]
+1. No serious objections.

File: chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs (hunk 2 above)
Comment (TheQuantumPhysicist) [reply chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs:22]: Nope. It's not possible. I already tried that yesterday. There's a destination that's pulled from the database through accounting.

File: clippy.toml [hunk 3]
```diff
@@ -1,3 +1,4 @@
 disallowed-methods = []
 disallowed-types = []
 enforced-import-renames = []
+large-error-threshold = 256
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply clippy.toml:4]: The problem with boxing errors is that you can't match with them anymore. There's a `box` qualifier for that but it's nightly. I don't see how else to do this.

[the remark this answers is not in the document]
Comment (iljakuklic) [reply clippy.toml:4]: Hmm ok.

File: chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs (hunk 2 above)
Comment (iljakuklic) [reply chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs:22]: Hmm I see so `Cow` at best, not worth it.

**Pull request merged**